File: tests/XRoadLib.Tests/XRoadMessageTest.fs
namespace XRoadLib.Tests

open FsUnit
open NUnit.Framework
open XRoadLib
open XRoadLib.Serialization
open System.Collections.Specialized
open System.IO
open System.Text

[<TestFixture>]
module ContentLengthTest =
    let writeContent (content, attachments) =
        use outputStream = new MemoryStream()
        use contentStream = new MemoryStream()
        content |> Option.iter (fun value ->
            let buf = Encoding.UTF8.GetBytes(value: string)
            contentStream.Write(buf, 0, buf.Length))
        use message = new XRoadMessage(contentStream)
        attachments |> List.iter message.AllAttachments.Add
        message.SaveTo(outputStream, (fun _ -> ()), (fun _ _ -> ()))
        message.ContentLength

    [<Test>]
    let ``write empty content`` () =
        (None, []) |> writeContent |> should equal 0

    [<Test>]
    let ``write content without attachments`` () =
        (Some("test"), []) |> writeContent |> should equal 4

    [<Test>]
    let ``write content with attachments`` () =
        use x = new XRoadAttachment(new MemoryStream([| 1uy; 2uy; 3uy; 4uy |]))
        use y = new XRoadAttachment(new MemoryStream([| 5uy; 6uy; 7uy; 8uy; 9uy |]))
        (Some("test2"), [x; y]) |> writeContent |> should equal 534

    [<Test>]
    let ``read empty content without attachments`` () =
        use stream = new MemoryStream()
        use reader = new XRoadMessageReader(stream, NameValueCollection(), Encoding.UTF8, Path.GetTempPath(), [Globals.XRoadProtocol20; Globals.XRoadProtocol31; Globals.XRoadProtocol40])
        use msg = new XRoadMessage()
        reader.Read(msg, false)
        msg.ContentLength |> should equal 0L

    [<Test>]
    let ``read content without attachments`` () =
        use stream = new MemoryStream()
        use streamWriter = new StreamWriter(stream, Encoding.UTF8)
        streamWriter.Write("""<Envelope xmlns="http://schemas.xmlsoap.org/soap/envelope/" xmlns:id="http://x-road.eu/xsd/identifiers" xmlns:repr="http://x-road.eu/xsd/representation.xsd">
  <Header xmlns:xrd="http://x-road.ee/xsd/x-road.xsd">
  </Header>
  <Body />
</Envelope>""")
        streamWriter.Flush()
        stream.Position <- 0L
        use reader = new XRoadMessageReader(stream, NameValueCollection(), Encoding.UTF8, Path.GetTempPath(), [Globals.XRoadProtocol20; Globals.XRoadProtocol31; Globals.XRoadProtocol40])
        use msg = new XRoadMessage()
        reader.Read(msg, false)
        msg.ContentLength |> should equal 254L

    [<Test>]
    let ``read content with attachments`` () =
        use stream = new MemoryStream()
        use streamWriter = new StreamWriter(stream, Encoding.UTF8)
        streamWriter.Write("""
--5e7a8827-5850-45be-9a1e-8bbf6aff5da7
Content-Type: text/xml; charset=UTF-8
Content-Transfer-Encoding: 8bit
Content-ID: <rQI0gpIFuQMxlrqBj3qHKw==>

<?xml version="1.0" encoding="utf-8"?>
<Envelope xmlns="http://schemas.xmlsoap.org/soap/envelope/" xmlns:id="http://x-road.eu/xsd/identifiers" xmlns:repr="http://x-road.eu/xsd/representation.xsd">
  <Header xmlns:xrd="http://x-road.ee/xsd/x-road.xsd">
  </Header>
  <Body />
</Envelope>

--5e7a8827-5850-45be-9a1e-8bbf6aff5da7
Content-Type: application/octet-stream
Content-Transfer-Encoding: binary
Content-ID: <CNbAWiFRKnmh3+udKo8mLw==>

proovikas
--5e7a8827-5850-45be-9a1e-8bbf6aff5da7
Content-Type: application/octet-stream
Content-Transfer-Encoding: binary
Content-ID: <qrOlKraewrdRAu86cFnqwg==>

testikas sisu
--5e7a8827-5850-45be-9a1e-8bbf6aff5da7--""")
        streamWriter.Flush()
        stream.Position <- 0L
        let headers = NameValueCollection()
        headers.Add("Content-Type", "multipart/related; type=\"application/xml+xop\"; start=\"rQI0gpIFuQMxlrqBj3qHKw==\"; boundary=\"5e7a8827-5850-45be-9a1e-8bbf6aff5da7\"")
        use reader = new XRoadMessageReader(stream, headers, Encoding.UTF8, Path.GetTempPath(), [Globals.XRoadProtocol20; Globals.XRoadProtocol31; Globals.XRoadProtocol40])
        use msg = new XRoadMessage()
        reader.Read(msg, false)
        msg.ContentLength |> should equal 838L

File: test/XRoadLib.Tests/Serialization/XRoadMessageTest.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using XRoadLib.Serialization;
using XRoadLib.Soap;
using Xunit;

namespace XRoadLib.Tests.Serialization
{
    public class XRoadMessageTest
    {
        private static readonly IMessageFormatter MessageFormatter = new SoapMessageFormatter();

        private readonly IServiceManager[] _serviceManagers = { Globals.ServiceManager };

        [Fact]
        public async Task CanWriteEmptyContent()
        {
            var contentLength = await WriteContentAsync();
            Assert.Equal(0L, contentLength);
        }

        [Fact]
        public async Task CanWriteContentWithoutAttachments()
        {
            var contentLength = await WriteContentAsync("test");
            Assert.Equal(4L, contentLength);
        }

        [Fact]
        public async Task CanWriteContentWithAttachments()
        {
            using var x = new XRoadAttachment(new MemoryStream(new byte[] { 1, 2, 3, 4 }));
            using var y = new XRoadAttachment(new MemoryStream(new byte[] { 5, 6, 7, 8, 9 }));
            var contentLength = await WriteContentAsync("test2", new[] { x, y });
            Assert.Equal(562L, contentLength);
        }

        [Fact]
        public async Task CanReadEmptyContentWithoutAttachments()
        {
            using var stream = new MemoryStream();
            using var reader = new XRoadMessageReader(new DataReader(stream), MessageFormatter, "text/xml; charset=UTF-8", Globals.StoragePath, _serviceManagers);
            using var message = new XRoadMessage();
            await reader.ReadAsync(message, false);
            Assert.Equal(0L, message.ContentLength);
        }

        [Fact]
        public async Task CanReadContentWithoutAttachments()
        {
            using var stream = new MemoryStream();
            using var streamWriter = new StreamWriter(stream, XRoadEncoding.Utf8);

            await streamWriter.WriteAsync("<Envelope xmlns=\"http://schemas.xmlsoap.org/soap/envelope/\" xmlns:id=\"http://x-road.eu/xsd/identifiers\" xmlns:repr=\"http://x-road.eu/xsd/representation.xsd\">\r\n");
            await streamWriter.WriteAsync("  <Header xmlns:xrd=\"http://x-road.ee/xsd/x-road.xsd\">\r\n");
            await streamWriter.WriteAsync("  </Header>\r\n");
            await streamWriter.WriteAsync("  <Body />\r\n");
            await streamWriter.WriteAsync("</Envelope>");
            await streamWriter.FlushAsync();

            stream.Position = 0L;

            using var reader = new XRoadMessageReader(new DataReader(stream), MessageFormatter, "text/xml; charset=UTF-8", Globals.StoragePath, _serviceManagers);
            using var message = new XRoadMessage();

            await reader.ReadAsync(message, false);
            Assert.Equal(251L, message.ContentLength);
        }

        [Fact]
        public async Task CanReadContentWithAttachments()
        {
            using var stream = new MemoryStream();
            using var streamWriter = new StreamWriter(stream, XRoadEncoding.Utf8);

            await streamWriter.WriteAsync("\r\n");
            await streamWriter.WriteAsync("--5e7a8827-5850-45be-9a1e-8bbf6aff5da7\r\n");
            await streamWriter.WriteAsync("Content-Type: text/xml; charset=UTF-8\r\n");
            await streamWriter.WriteAsync("Content-Transfer-Encoding: 8bit\r\n");
            await streamWriter.WriteAsync("Content-ID: <rQI0gpIFuQMxlrqBj3qHKw==>\r\n");
            await streamWriter.WriteAsync("\r\n");
            await streamWriter.WriteAsync("<?xml version=\"1.0\" encoding=\"utf-8\"?>\r\n");
            await streamWriter.WriteAsync("<Envelope xmlns=\"http://schemas.xmlsoap.org/soap/envelope/\" xmlns:id=\"http://x-road.eu/xsd/identifiers\" xmlns:repr=\"http://x-road.eu/xsd/representation.xsd\">\r\n");
            await streamWriter.WriteAsync("  <Header xmlns:xrd=\"http://x-road.ee/xsd/x-road.xsd\">\r\n");
            await streamWriter.WriteAsync("  </Header>\r\n");
            await streamWriter.WriteAsync("  <Body />\r\n");
            await streamWriter.WriteAsync("</Envelope>\r\n");
            await streamWriter.WriteAsync("\r\n");
            await streamWriter.WriteAsync("--5e7a8827-5850-45be-9a1e-8bbf6aff5da7\r\n");
            await streamWriter.WriteAsync("Content-Type: application/octet-stream\r\n");
            await streamWriter.WriteAsync("Content-Transfer-Encoding: binary\r\n");
            await streamWriter.WriteAsync("Content-ID: <CNbAWiFRKnmh3+udKo8mLw==>\r\n");
            await streamWriter.WriteAsync("\r\n");
            await streamWriter.WriteAsync("proovikas\r\n");
            await streamWriter.WriteAsync("--5e7a8827-5850-45be-9a1e-8bbf6aff5da7\r\n");
            await streamWriter.WriteAsync("Content-Type: application/octet-stream\r\n");
            await streamWriter.WriteAsync("Content-Transfer-Encoding: binary\r\n");
            await streamWriter.WriteAsync("Content-ID: <qrOlKraewrdRAu86cFnqwg==>\r\n");
            await streamWriter.WriteAsync("\r\n");
            await streamWriter.WriteAsync("testikas sisu\r\n");
            await streamWriter.WriteAsync("--5e7a8827-5850-45be-9a1e-8bbf6aff5da7--");
            await streamWriter.FlushAsync();

            const string contentTypeHeader = "multipart/related; type=\"application/xml+xop\"; start=\"rQI0gpIFuQMxlrqBj3qHKw==\"; boundary=\"5e7a8827-5850-45be-9a1e-8bbf6aff5da7\"";

            stream.Position = 0L;

            using var reader = new XRoadMessageReader(new DataReader(stream), MessageFormatter, contentTypeHeader, Globals.StoragePath, _serviceManagers);
            using var message = new XRoadMessage();

            await reader.ReadAsync(message, false);
            Assert.Equal(835L, message.ContentLength);
        }

        private static async Task<long> WriteContentAsync(string content = null, IEnumerable<XRoadAttachment> attachments = null)
        {
            var messageFormatter = new SoapMessageFormatter();

            using var outputStream = new MemoryStream();
            using var contentStream = new MemoryStream();

            if (content != null)
            {
                var buffer = XRoadEncoding.Utf8.GetBytes(content);
                contentStream.Write(buffer, 0, buffer.Length);
            }

            using var message = new XRoadMessage(contentStream);

            (attachments ?? Enumerable.Empty<XRoadAttachment>()).ToList().ForEach(message.AllAttachments.Add);
            await message.SaveToAsync(outputStream, x => {}, (k, v) => {}, messageFormatter);

            return message.ContentLength;
        }
    }
}